tabular/src/autogluon/tabular/configs/config_helper.py

```diff
@@ -0,0 +1,357 @@
+        Appends DatetimeFeatureGenerator() to the generator group.
+        """
+        self.config['enable_datetime_features'] = value
+        return self
+
+    def enable_text_special_features(self, value: bool = True) -> FeatureGeneratorBuilder:
+        """
+        Whether to use 'object' features identified as 'text' features to generate 'text_special' features such as word count, capital letter ratio, and symbol counts.
+        Appends TextSpecialFeatureGenerator() to the generator group.
+        """
+        self.config['enable_text_special_features'] = value
+        return self
+
+    def enable_text_ngram_features(self, value: bool = True) -> FeatureGeneratorBuilder:
+        """
+        Whether to use 'object' features identified as 'text' features to generate 'text_ngram' features.
+        Appends TextNgramFeatureGenerator(vectorizer=vectorizer) to the generator group.
+        """
+        self.config['enable_text_ngram_features'] = value
+        return self
+
+    def enable_raw_text_features(self, value: bool = True) -> FeatureGeneratorBuilder:
+        """
+        Whether to keep the raw text features.
+        Appends IdentityFeatureGenerator(infer_features_in_args=dict(required_special_types=['text'])) to the generator group.
+        """
+        self.config['enable_raw_text_features'] = value
+        return self
+
+    def enable_vision_features(self, value: bool = True) -> FeatureGeneratorBuilder:
+        """
+        [Experimental]
+        Whether to keep 'object' features identified as 'image_path' special type. Features of this form should have a string path to an image file as their value.
+        Only vision models can leverage these features, and these features will not be treated as categorical.
+        Note: 'image_path' features will not be automatically inferred. These features must be explicitly specified as such in a custom FeatureMetadata object.
+        Note: It is recommended that the string paths use absolute paths rather than relative, as it will likely be more stable.
+        """
+        self.config['enable_vision_features'] = value
+        return self
+
+    def vectorizer(self, value: BaseEstimator) -> FeatureGeneratorBuilder:
+        """
+        sklearn CountVectorizer object to use in TextNgramFeatureGenerator.
+        Only used if `enable_text_ngram_features=True`.
+        """
+        self.config['vectorizer'] = value
+        return self
+
+    def text_ngram_params(self, value: bool = True) -> FeatureGeneratorBuilder:
+        """
+        Appends TextNgramFeatureGenerator(vectorizer=vectorizer, text_ngram_params) to the generator group. See text_ngram.py for valid parameters.
+        """
+        self.config['text_ngram_params'] = value
+        return self
+
+    def build(self) -> ConfigBuilder:
+        self.parent.config['feature_generator'] = AutoMLPipelineFeatureGenerator(**self.config)
+        return self.parent
+
+
+class ConfigBuilder:
```

Comment:
[Major Question]

Are there examples of ML packages that use builder patterns? I'd like to at least compare our implementation to something widely used. If no ML packages use builder patterns, then that would also be useful information.

Reply:
[Method chaining](https://en.wikipedia.org/wiki/Method_chaining) is heavily used in pandas. Builder pattern is used in `.groupby` - it returns an object methods you can do on the group operation. `pyplot` is effectively a massive builder because the actual chart built on save methods or when the chart has to be displayed; though it's not things in a chaining session, more like a context, which is mutated by different calls. 

Effectively the best time to use builders are: 1) complex object hierarchy creation/management (like pyplot/urllib) or 2) convenience: massive number properties.

Reply:
I'm a bit concerned on the value add here vs the cost of maintaining it. I don't think ML users are familiar with builder patterns, so I'm struggling to see a large portion of users preferring this API to the general TabularPredictor API. We can certainly give it a try, but unsure how much it will be used.

Reply:
I think we should review what should be part of core basic APIs `.fit` and what parts are syntactic sugar on top of it and should not result into adding new things to core APIs. It's much easier to handle parameters configuration and interaction code when it is separated from ML code.

Reply:
Definitely worth a discussion on how to simplify APIs. Tough part will be that it will be breaking (we could do a deprecation warning though)